packages/react-hooks/src/index.tsx

```diff
@@ -0,0 +1,277 @@
+    proofs: [],
+    loading: true,
+  })
+  const [basicMessageState, setBasicMessageState] = useState<{
+    basicMessages: BasicMessageRecord[] | []
+    loading: boolean
+  }>({ basicMessages: [], loading: true })
+
+  useEffect(() => {
+      setInitialState()
+  }, [agent])
+
+  const setInitialState = async () => {
+    try {
+      if(agent) {
+        const connections = await agent.connections.getAll()
+        const credentials = await agent.credentials.getAll()
+        const proofs = await agent.proofs.getAll()
+        const basicMessages = await agent.basicMessages.findAllByQuery({})
+
+        setAgentState({ agent, loading: false })
+        setConnectionState({ connections, loading: false })
+        setCredentialState({ credentials, loading: false })
+        setProofState({ proofs, loading: false })
+        setBasicMessageState({ basicMessages, loading: false })
+      }
+    } catch (e) {
+      console.error(TAG, 'ERROR IN SET_INITIAL_STATE:', e)
+    }
+  }
+
+  useEffect(() => {
+    if (!connectionState.loading) {
+      const listener = (event: ConnectionStateChangedEvent) => {
+        console.debug(TAG, 'Connection Event', event)
+
+        let newConnectionsState = connectionState.connections
+        const index = newConnectionsState.findIndex((connection) => connection.id === event.payload.connectionRecord.id)
+        if (index > -1) {
+          newConnectionsState[index] = event.payload.connectionRecord
+        } else {
+          newConnectionsState = [...newConnectionsState, event.payload.connectionRecord]
+        }
+
+        setConnectionState({
+          loading: connectionState.loading,
+          connections: newConnectionsState,
+        })
+      }
+      agentState.agent?.events.on(ConnectionEventTypes.ConnectionStateChanged, listener)
+
+      return () => {
+        agentState.agent?.events.off(ConnectionEventTypes.ConnectionStateChanged, listener)
+      }
+    }
+  }, [connectionState])
+
+  useEffect(() => {
+    if (!credentialState.loading) {
+      const listener = async (event: CredentialStateChangedEvent) => {
+        console.debug(TAG, 'Credential Event', event)
+
+        let newCredentialsState = credentialState.credentials
+        const index = newCredentialsState.findIndex((credential) => credential.id === event.payload.credentialRecord.id)
+        if (index > -1) {
+          newCredentialsState[index] = event.payload.credentialRecord
+        } else {
+          newCredentialsState = [...newCredentialsState, event.payload.credentialRecord]
+        }
+
+        setCredentialState({
+          loading: credentialState.loading,
+          credentials: newCredentialsState,
+        })
+      }
+
+      agentState.agent?.events.on(CredentialEventTypes.CredentialStateChanged, listener)
+
+      return () => {
+        agentState.agent?.events.off(CredentialEventTypes.CredentialStateChanged, listener)
+      }
+    }
```

Comment:
why are all listeners removed during loading? This means if an event is fired during loading we won't get it in our local state

Reply:
> Why do we not add the loaders when loading=true?

Since we're loading the `agent` asynchronously (which differs from the redux-store package _I think_), then there's a period of time in the aries-hooks package that `agent` is undefined, so if we try and register the listener while loading = `true` then we'll see an error trying to register the listener using something undefined.
AFAIK the cleanup function isn't triggered immediately, only when the useEffect is being ran again or the component is being unmounted. I could be wrong in that statement though.

I am all ears if I am missing something stupid here! :)

Reply:
No worries--totally could see your concern if you didn't see how the loading param was working.